Experimental Workload: Responsive Design

issackjohn opened the pull request:
This pull request introduces a new experimental responsive design workload, featuring a single-page cooking website built with Lit and Tailwind CSS.

The hosted version can be found [here](https://neon-gumdrop-d49b03.netlify.app/).
[Quick link to dev mode with workload selected.
](https://neon-gumdrop-d49b03.netlify.app/?developerMode=&suites=Responsive-Design)
Please look at the [README](https://github.com/WebKit/Speedometer/pull/453/files#diff-83d565323d6c3520a4207b0f9b17122f2c51a3a16238c868e7e53effacfffc9f) in `experimental/responsive-design` for more information about the code and how it is organized.

Not included in this PR:
- Support for different languages

issackjohn commented:
Adding reviewers for visibility, feel free to take a look :)

rniwa commented:
It seems to me that the test doesn't really take the advantage of responsive design since iframe's size is always same. Maybe we should add steps to resize iframe so that responsive design comes into play?

issackjohn commented:
> It seems to me that the test doesn't really take the advantage of responsive design since iframe's size is always same. Maybe we should add steps to resize iframe so that responsive design comes into play?

Currently, we are resizing the the iframe with id `#content-iframe` in the [ReduceWidthIn5Steps](https://github.com/WebKit/Speedometer/pull/453/files#diff-2c1c126bd9bb884254da07a36c11788febf37ba69b0afcb3b7a735ca4d03b287R1122-R1130) & [IncreaseWidthIn5Steps](https://github.com/WebKit/Speedometer/pull/453/files#diff-2c1c126bd9bb884254da07a36c11788febf37ba69b0afcb3b7a735ca4d03b287R1143-R1151) steps. Do you mean that we should be resizing a different iframe?

rniwa commented:
> > It seems to me that the test doesn't really take the advantage of responsive design since iframe's size is always same. Maybe we should add steps to resize iframe so that responsive design comes into play?
> 
> Currently, we are resizing the the iframe with id `#content-iframe` in the [ReduceWidthIn5Steps](https://github.com/WebKit/Speedometer/pull/453/files#diff-2c1c126bd9bb884254da07a36c11788febf37ba69b0afcb3b7a735ca4d03b287R1122-R1130) & [IncreaseWidthIn5Steps](https://github.com/WebKit/Speedometer/pull/453/files#diff-2c1c126bd9bb884254da07a36c11788febf37ba69b0afcb3b7a735ca4d03b287R1143-R1151) steps. Do you mean that we should be resizing a different iframe

Oh, I see. We're resizing the inner iframe. I guess that works.

issackjohn commented:
Gentle ping 😄

camillobruni commented:
I still see rather short run times for individual steps (close to 1.6ms in chrome canary) which might introduce a bit too much noise in the long run.

```
Responsive-Design
Responsive-Design/LoadChatAndExpandRecipe	5.85	±	0.46 (7.9%)	ms
Responsive-Design/ReduceWidthIn5Steps  12.96	±	0.45 (3.5%)	ms
Responsive-Design/ScrollToChatAndSendMessage	1.68	±	0.07 (4.4%)	ms
Responsive-Design/IncreaseWidthIn5Steps	11.37	±	0.28 (2.5%)	
```


ScrollToChatAndSendMessage and (and especially it's sub-items) will show some aliasing from timer resolution which is ~ 0.1ms on chrome.
In general, i'd bump the step durations by 1.5x or 2x even to be on par with other workloads.

camillobruni commented on experimental/responsive-design/index.html:
```diff
@@ -0,0 +1,11 @@
+        <title>Cooking with Lit and Tailwind</title>
+    </head>
+    <body style="overflow: hidden; margin: 0">
+        <iframe title="Cooking Website with Recipes" id="content-iframe" src="iframe.html" width="800px" height="600px" style="border: none"></iframe>
```
Can we start off with 100% width/height as the maximum size (the step-resize sounds ok to be done with explicit sizes again)?
Albeit not fully exposed, we can set a custom viewport size in the params.mjs and it's used in the runner. 
This workload would currently not support this directly.

issackjohn replied:
Sounds good to me

issackjohn commented:
> I still see rather short run times for individual steps (close to 1.6ms in chrome canary) which might introduce a bit too much noise in the long run.
> 
> ```
> Responsive-Design
> Responsive-Design/LoadChatAndExpandRecipe	5.85	±	0.46 (7.9%)	ms
> Responsive-Design/ReduceWidthIn5Steps  12.96	±	0.45 (3.5%)	ms
> Responsive-Design/ScrollToChatAndSendMessage	1.68	±	0.07 (4.4%)	ms
> Responsive-Design/IncreaseWidthIn5Steps	11.37	±	0.28 (2.5%)	
> ```
> 
> ScrollToChatAndSendMessage and (and especially it's sub-items) will show some aliasing from timer resolution which is ~ 0.1ms on chrome. In general, i'd bump the step durations by 1.5x or 2x even to be on par with other workloads.

To bump up the step durations, I am thinking of
1. Expanding all the recipes as part of `LoadChatAndExpandRecipe`
2. Collapsing all the recipes as part of `ScrollToChatAndSendMessage`
2.1 Sending a few more messages in the chat instead of just one. 

currently no proposed changes for `ReduceWidthIn5Steps` & `IncreaseWidthIn5Steps`

flashdesignory commented:
Some visual questions - these might not be relevant to the tests:

Is this spacing as expected?
<img width="1405" alt="Screenshot 2024-12-09 at 10 08 00 PM" src="https://github.com/user-attachments/assets/bb263727-5753-4cbd-a965-d9b80870765d">

The collapsed grid looks fine, but once expanded, the item moves to the next row - is that expected?
collapsed
<img width="1372" alt="Screenshot 2024-12-09 at 10 08 09 PM" src="https://github.com/user-attachments/assets/243ef36a-c7d3-423a-bf8f-69bb667f3d7b">

expanded
<img width="1407" alt="Screenshot 2024-12-09 at 10 08 15 PM" src="https://github.com/user-attachments/assets/90cb51ca-0c70-4458-80d7-cbc84d934d0e">

The video tiles don't work - those are just placeholders? 
<img width="1366" alt="Screenshot 2024-12-09 at 10 08 34 PM" src="https://github.com/user-attachments/assets/775c80eb-cf40-4d3c-9ee2-ad9d3f6ebf5d">

issackjohn commented:
> Some visual questions - these might not be relevant to the tests:
> 
> Is this spacing as expected? 

I'll add a fix for that. The webpage was designed to be viewed in 800x600, as that's how it will appear within the benchmark.

> The collapsed grid looks fine, but once expanded, the item moves to the next row - is that expected? collapsed 

Yes, that is expected. Do we prefer a different behavior?

> The video tiles don't work - those are just placeholders? 

Yes, they are just placeholders. We intended the video elements to be there but did not plan for them to play any actual videos.

flashdesignory commented:
> > Some visual questions - these might not be relevant to the tests:
> > Is this spacing as expected?
> 
> I'll add a fix for that. The webpage was designed to be viewed in 800x600, as that's how it will appear within the benchmark.

 - Thanks - I'd vote for supporting larger viewport sizes. This will ensure that the workload looks usable at larger viewports, in case we want to change what's supported in Speedometer later on. 

> 
> > The collapsed grid looks fine, but once expanded, the item moves to the next row - is that expected? collapsed
> 
> Yes, that is expected. Do we prefer a different behavior?

I guess it makes sense for the tests, since it shifts elements on the page. That said, it's not something you'd see on the web (hopefully). I'd expected a modal or something like that instead.

> 
> > The video tiles don't work - those are just placeholders?
> 
> Yes, they are just placeholders. We intended the video elements to be there but did not plan for them to play any actual videos.

Thanks for clarifying.

flashdesignory commented:
@issackjohn - maybe rebase / merge main, to get the updated test runner in there. That should ensure the checks don't fail.

camillobruni commented:
Let's aim for > 5ms for each measured subtest (I'd still aim for roughly 10ms if possible). Does this sound reasonable?

issackjohn commented:
> @issackjohn - maybe rebase / merge main, to get the updated test runner in there. That should ensure the checks don't fail.

Sounds good.

issackjohn commented:
> Let's aim for > 5ms for each measured subtest (I'd still aim for roughly 10ms if possible). Does this sound reasonable?

Sounds reasonable to me, I'll add some more actions.

issackjohn commented:
Thanks! Looking forward to addressing anymore feedback from folks.

julienw commented on resources/benchmark-runner.mjs:
```diff
@@ -62,13 +62,13 @@ class Page {
-    querySelector(selector, path = []) {
-        const lookupStartNode = this._frame.contentDocument;
+    querySelector(selector, path = [], inIframe = false) {
+        const lookupStartNode = inIframe ? this._frame.contentDocument : this._frame.contentDocument;
```
The 2 parts of the ternary are identical, is that expected?

julienw commented on resources/tests.mjs:
```diff
@@ -1113,6 +1113,90 @@ Suites.push({
+
+            widths.forEach((width) => {
+                iframeElement.setWidth(`${width}px`);
+                page.layout();
```
I think the layout doesn't work, possibly because the resize observer is async? 
I think you should try doing several steps instead of just one, so that the async part is properly captured.

julienw commented on resources/tests.mjs:
```diff
@@ -1113,6 +1113,90 @@ Suites.push({
+            const widths = [560, 640, 704, 768, 800];
+
+            widths.forEach((width) => {
+                iframeElement.setWidth(`${width}px`);
```
same here

julienw reviewed (dismissed):
Ths looks pretty good to me!

I see that when running `npm run dev` the dist files get overridden. Do you know if we can make it so the dev files are written in a different location, to minimize risks of errors in the future?

Otherwise my main issue is that the width setting in `ReduceWidthIn5Steps` and `IncreaseWidthIn5Steps` don't seem to be captured properly, at least in Firefox on my current machine (which is a bit slow). IMO this needs to be fixed, but this can be done in a follow-up.

Then I have a few more minor comments here and there. I haven't looked very closely at the tailwind code because I'm not familiar with it.
Also I wonder if we could include `matchMedia` somewhere.

Thanks!

Please add new changes in a separate commit.

camillobruni commented on resources/tests.mjs:
```diff
@@ -1113,6 +1113,90 @@ Suites.push({
+            }
+
+            const element = iframeElement.querySelectorInShadowRoot("#chat-window", ["cooking-app", "chat-window"]);
+            element.scrollIntoView();
```
We should probably explicitly force the "instant" behavior. If we accidentally get "smooth" it will likely be not measured correctly.

rniwa commented on experimental/responsive-design/src/input.css:
```diff
@@ -0,0 +1,29 @@
+
+@layer utilities {
+    .content-auto {
+        content-visibility: auto;
```
Is `content-visibility` widely supported across browsers and used on popular sites?

julienw replied:
Actually it seems supported partially in Safari.

From the chromestatus page, can we see which different values are used and not just the property?

issackjohn replied:
Took a look and it seems like you can.
<img width="933" height="192" alt="image" src="https://github.com/user-attachments/assets/2d495ca9-f755-4a37-b96a-e806132c7029" />
<img width="324" height="245" alt="image" src="https://github.com/user-attachments/assets/1c75c50f-dc86-456e-8c70-4c9318b5478b" />

julienw replied:
"auto" seems to be supported as of Safari 26 (2025-09-15⁩) so it's pretty new. Do we want to keep it?

issackjohn replied:
We prefer to keep it. I'll let others comment as well. @rniwa

rniwa commented on experimental/responsive-design/src/lib/components/information-window.js:
```diff
@@ -0,0 +1,116 @@
+
+    connectedCallback() {
+        super.connectedCallback();
+        adoptStyles(this.shadowRoot, [chatWindowStyles]);
```
Don't we want / need to re-install resize observer here?
Same as other classes that use resize observer and disconnect in `disconnectedCallback`.

issackjohn commented on experimental/responsive-design/.prettierrc:
```diff
@@ -0,0 +1,13 @@
+    "trailingComma": "es5",
+    "useTabs": false,
+    "plugins": ["prettier-plugin-tailwindcss"]
+}
```
Using project wide prettier now.

issackjohn commented:
> Then I have a few more minor comments here and there. I haven't looked very closely at the tailwind code because I'm not familiar with it. Also I wonder if we could include `matchMedia` somewhere.
> 

Sounds good. I can look into incorporating matchMedia as a follow-up and see where it might make sense to add.

issackjohn commented:
> I see some lazy recalc-style happening outside the measured scope in chrome. (around 3ms) I'm not sure how we'd fully control this. From Devtools it clearly looks like work that was scheduled in the async phase. I've tried to see whether this happens in Firefox as well, but couldn't spot this immediately. Might be worthwhile checking on Safari as well.
> 
> <img alt="Screenshot 2025-07-01 at 14 28 55" width="526" src="https://private-user-images.githubusercontent.com/129550/461046329-05c2e8a3-fb6c-4eca-89d2-3775ddebb9f5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTMxMTEyMzUsIm5iZiI6MTc1MzExMDkzNSwicGF0aCI6Ii8xMjk1NTAvNDYxMDQ2MzI5LTA1YzJlOGEzLWZiNmMtNGVjYS04OWQyLTM3NzVkZGViYjlmNS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNzIxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDcyMVQxNTE1MzVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01ZmIzZjI5YjgxODVkNTk1YzI3YjgxNjdiZjM2Nzk1MjEzYTU2NjQ0Y2I0NzBlNjY3M2MxYzQyNDUxZGZhNzQ1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.hOI6TpWop6a42HiwhOSnQjm-3t2awygkBnQvD7h-tYo">

Now capturing in measured time on edge/chrome
- `ReduceWidthIn5Steps` and `IncreaseWidthIn5Steps` when using `getBoundingRectAndElementFromPoint` 
- `LoadChatAndExpandRecipes` when using `getBoundingClientRect`.

Working on properly capturing the other steps when using `getBoundingClientRect`.

issackjohn commented:
> > I see some lazy recalc-style happening outside the measured scope in chrome. (around 3ms) I'm not sure how we'd fully control this. From Devtools it clearly looks like work that was scheduled in the async phase. I've tried to see whether this happens in Firefox as well, but couldn't spot this immediately. Might be worthwhile checking on Safari as well.
> > <img alt="Screenshot 2025-07-01 at 14 28 55" width="526" src="https://private-user-images.githubusercontent.com/129550/461046329-05c2e8a3-fb6c-4eca-89d2-3775ddebb9f5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTMxMTEyMzUsIm5iZiI6MTc1MzExMDkzNSwicGF0aCI6Ii8xMjk1NTAvNDYxMDQ2MzI5LTA1YzJlOGEzLWZiNmMtNGVjYS04OWQyLTM3NzVkZGViYjlmNS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNzIxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDcyMVQxNTE1MzVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01ZmIzZjI5YjgxODVkNTk1YzI3YjgxNjdiZjM2Nzk1MjEzYTU2NjQ0Y2I0NzBlNjY3M2MxYzQyNDUxZGZhNzQ1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.hOI6TpWop6a42HiwhOSnQjm-3t2awygkBnQvD7h-tYo">
> 
> Now capturing in measured time on edge/chrome
> 
> * `ReduceWidthIn5Steps` and `IncreaseWidthIn5Steps` when using `getBoundingRectAndElementFromPoint`
> * `LoadChatAndExpandRecipes` when using `getBoundingClientRect`.
> 
> Working on properly capturing the other steps when using `getBoundingClientRect`.

All steps should now be properly captured. 

I've observed that some work occasionally occurs outside the measured time following the `ScrollToChatAndSendMessages` step. From my understanding, this shouldn't be an issue. However, do we believe this activity should be included in the measured time? If we decide it's worth tracking, do you have any ideas of how we could capture that work?
<img width="1433" height="741" alt="image" src="https://github.com/user-attachments/assets/2a4ad8a5-558a-4e01-a5a5-fd4d95e4c570" />

issackjohn commented:
@bgrins, @camillobruni, @rniwa, Could I get someone to take a look at the `matchMedia` implementation? Most of it is in `experimental/responsive-design/src/lib/components/recipe-grid.js` and used to switch the recipe cards to a compact version. 

I'm still working on correctly capturing the other work caused by things like ResizeObserver within the measured time.

issackjohn commented:
Gentle Ping. :)

issackjohn commented:
Created https://github.com/WebKit/Speedometer/issues/525 for safari CI run end2end Test failure.

issackjohn commented:
Gentle ping :)

camillobruni commented on experimental/responsive-design/src/lib/components/recipe-carousel.js:
```diff
@@ -0,0 +1,121 @@
+        this._currentIndex = 0;
+        this._carouselWidth = 0;
+        // ResizeObserver is used primarily to exercise this API as part of the benchmark.
+        this._resizeObserver = new ResizeObserver((entries) => {
```
maybe it would be nice to track the number of resize events we get so we can verify that this is concistent across browsers (console.warn  + a strict unittest or so)

issackjohn replied:
I will do this in a follow up, thanks!

camillobruni commented:
Generally LGTM especially for experimental.

I'd love to land this so we can start tracking in and see how well it behaves.

issackjohn committed:
Add responsive design benchmark

issackjohn committed:
Keep Responsive-Design content visibility wait live (#51)

Safari stable can overscroll the chat element in this workload, tracked in https://github.com/WebKit/Speedometer/issues/525. When that happens, the video grid's contentvisibilityautostatechange event may never arrive.
